New coding convention: use lemma/theorem and tactics when proving propositions (#240)
runtimeverification/vlsm: CODING_CONVENTIONS.md

  set_union (lv_sent_observations s target) (lv_received_observations s target).
```

> - do not use definitions to prove propositions; use lemma/theorem and tactics instead

Reviewer: I think this is quite far in precision and accuracy from what @bmmoore and I discussed elsewhere. Specifically:
- we don't really talk about "propositions", but about terms that live in the sort `Prop` (which have Coq's weak form of proof irrelevance)
- we don't necessarily want people to use tactics beyond a single call to `exact`, and this is only to get the constants to become opaque

So, for example, one would write the "recommended" version as:
```coq
Proof. exact (map_option_app _). Qed.
```

Author: By "proposition" I meant a thing living in `Prop`, I can reword it if you want.

I don't understand your second point. We don't want people to use `exact`, do we? We don't use it when a subgoal can be proved by giving a term, so why should we if the entire lemma can be proved by giving a term? Also, what's the point of writing the term explicitly if we don't want to rely on the exact term and make it opaque?

Reviewer: The point is that writing explicit terms is the most stable and reliable way to write Coq constants. `exact` is precisely the right tactic to do this. Using tactics besides `exact` when they aren't needed leads to much harder-to-maintain code. Our goal here is to enable having opaque constant bodies even when using `Lemma`, and `Proof. exact <term>. Qed.` is the best way to accomplish this. I will request changes to any PR that changes the body of an existing `Definition` living in `Prop` to anything more loose than `exact`.

Reviewer 2: ```suggestion
- To give a proof term for something in `Prop`, prefer a `Proof` with `explicit` over a `Definition` body (with `:=`). because it can be unclear whether the goal was just to give an explicit proof term, or if transparency was also intentional. To give an explicit term instead use a `Proof` with `exact` use because it is not clear whether transparency is intended or a side effect
```

> - When giving an explicit proof term for things that live in `Prop`, avoid using `Definition` with `:=`, because it is unclear whether the point is just to give the proof term or whether transparency also matters. Instead use `Lemma` and provide the proof term using the `exact` tactic, while marking transparency with `Qed` or `Defined`.

Reviewer: ```suggestion
- When explicitly defining a constant that lives in `Prop`, avoid using `Definition` with `:=`. Instead use `Lemma` (or `Theorem`, `Proposition`, etc.) and provide the constant term body using the `exact` tactic, while indicating the opacity with `Qed` or `Defined`.
```

Author: Changed.

Not recommended:
```coq
Definition finite_trace_partial_map_app
  : forall l1 l2, finite_trace_partial_map (l1 ++ l2) =
    finite_trace_partial_map l1 ++ finite_trace_partial_map l2
  := map_option_app _.
```

Recommended:
```coq
Lemma finite_trace_partial_map_app :
  forall l1 l2 : list (transition_item TX),
    finite_trace_partial_map (l1 ++ l2) =
    finite_trace_partial_map l1 ++ finite_trace_partial_map l2.

> Proof.
>   exact (map_option_app _).
> Qed.

Reviewer: ```suggestion
Proof. exact (map_option_app _). Qed.
```

Author: Changed.
